### PR TITLE
Added functional tests for JsonPatch

### PR DESCRIPTION
**Pull request by kichalla**

### CODE REVIEW

**Comment by kichalla** [comment]
ping

**Comment by kichalla** [comment]
gentle ping

File: test/WebSites/FormatterWebSite/Controllers/JsonPatchController.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+namespace FormatterWebSite.Controllers
+{
+    [Route("jsonpatch/[action]")]
+    public class JsonPatchController : Controller
+    {
+        [HttpPatch]
+        public IActionResult PatchProduct([FromBody] JsonPatchDocument<Product> patchDoc)
+        {
+            if (patchDoc != null)
```
Comment (rynowak) [inline test/WebSites/FormatterWebSite/Controllers/JsonPatchController.cs]: This null check isn't needed.

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/JsonPatchInputFormatterTest.cs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            var body = await response.Content.ReadAsStringAsync();
+            var product = JsonConvert.DeserializeObject<Product>(body);
+            Assert.NotNull(product);
+            Assert.NotNull(product.Reviews);
+            Assert.Equal(2, product.Reviews.Count);
+            Assert.Equal(0, product.Reviews[0].Rating);
+        }
+
+        private HttpRequestMessage GetPatchRequest(string body)
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/JsonPatchInputFormatterTest.cs]: We should include one case where there are some errors as a result of the operation. This has way more integration risk than the successful operations.

**Comment by kichalla** [comment]
🆙 📅

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/JsonPatchInputFormatterTest.cs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            // Assert
+            Assert.Equal(HttpStatusCode.BadRequest, response.StatusCode);
+            var body = await response.Content.ReadAsStringAsync();
+            Assert.Equal(
+                "{\"Product\":[\"The value '{\r\n  \"Rating\": \"not-a-double\"\r\n}' is invalid for target location.\"]}",
+                body);
+        }
+
+        [Fact]
+        public async Task InvalidJsonPatch_RequestEnvelope_ResultsInBadRequest()
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/JsonPatchInputFormatterTest.cs]: I think that this name is confusing. I'd just mention here, either by improving the name or adding a comment, that the `add` operation requires a value and this doesn't provide it.

**Pull request merged**